blockchain/vm/evm.go

```diff
@@ -538,6 +539,21 @@ func (evm *EVM) CreateWithAddress(caller types.ContractRef, code []byte, gas uin
 	return evm.create(caller, codeAndHash, gas, value, contractAddr, humanReadable, codeFormat)
 }
 
+func (evm *EVM) GetPrecompiledContractMap(caller common.Address) map[common.Address]PrecompiledContract {
+	switch {
+	case evm.chainRules.IsIstanbul:
+		// if contract(caller) is deployed before istanbul, use the old precompiled contract set (use constantinople)
+		//      (gas price policy also follows constantinople rules)
+		// Without these lines, contracts that are deployed before istanbul and uses 0x09-0x0b won't work properly.
+		if evm.StateDB.GetCodeFormat(caller).IsDeployedAfterIstanbulHF() {
+			return PrecompiledContractsIstanbul
+		}
+		return PrecompiledContractsConstantinople
+	default:
+		return PrecompiledContractsConstantinople
+	}
```

Comment:
Doesn't this logic apply for after versions, too?
```
	// if contract(caller) is deployed before istanbul, use the old precompiled contract set (use constantinople)
	//      (gas price policy also follows constantinople rules)
	// Without these lines, contracts that are deployed before istanbul and uses 0x09-0x0b won't work properly.
	if !evm.StateDB.GetCodeFormat(caller).IsDeployedAfterIstanbulHF() {
			return PrecompiledContractsConstantinople
	}
	switch {
	case evm.chainRules.IsPetersburg: // this line is an example for next hard fork
		return PrecompiledContractsPetersburg
	case evm.chainRules.IsIstanbul:
		return PrecompiledContractsIstanbul
	default:
		return PrecompiledContractsConstantinople
	}
```

Reply:
@winnie-byun Thanks for picking up the unexpected part! Before, it does not consider another hardforks. Fixed the function at https://github.com/klaytn/klaytn/pull/912/commits/ad9898839583ea58277714efc58d0ae3012cc37c. Please take a look.